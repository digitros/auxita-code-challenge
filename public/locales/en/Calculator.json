{
  "BeforeCalc": "Here you will see your health calculation.",
  "Class": "Classification",
  "Date": "Date",
  "SeveDecreased": "Severely Decreased",
  "DropPercentage": "Drop Percentage",
  "Failure": "Kidney Failure",
  "ImportantDrops": "Important Drops",
  "IncorrectFormat": "Incorrect Format",
  "Input": "Please write/paste your data in following format",
  "LastDate": "Last Date",
  "LastReading": "Last Reading",
  "MildModerate": "Mild to Moderate",
  "MildyDecreased": "Mildy Decreased",
  "ModSevere": "Moderate to Severe",
  "Normal": "Normal",
  "NoStage": "No Hypertension",
  "Send": "Send",
  "Stage1": "Stage 1",
  "Stage2": "Stage 2",
  "Stage3": "Stage 3"
}
