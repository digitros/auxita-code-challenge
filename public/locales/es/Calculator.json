{
  "BeforeCalc": "Aquí verá el cálculo de su salud.",
  "Class": "Clasificación",
  "Date": "Fecha",
  "SeveDecreased": "Gravemente disminuido",
  "DropPercentage": "Porcentaje de caida",
  "Failure": "Falla del riñón",
  "ImportantDrops": "Caidas importantes",
  "IncorrectFormat": "Formato incorrecto",
  "Input": "Por favor escribe/pega tus datos en el siguiente formato",
  "LastDate": "Última fecha",
  "LastReading": "Última lectura",
  "MildModerate": "Leve a Moderado",
  "MildyDecreased": "Disminución Leve",
  "ModSevere": "Moderado a Severo",
  "Normal": "Normal",
  "NoStage": "Sin Hipertensión",
  "Send": "Enviar",
  "Stage1": "Etapa 1",
  "Stage2": "Etapa 2",
  "Stage3": "Etapa 3"
}
