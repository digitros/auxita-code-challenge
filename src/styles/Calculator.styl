.tools
    display: flex
    justify-content: space-evenly
    flex-direction: column
    &-title
        width: 100%
        display: flex
        justify-content: center
        flex-direction: column
        align-items: center
    &-main
        display: flex
        justify-content: space-evenly
        width: 100%

.calculator
    display: flex
    flex-direction: column
    justify-content: center
    align-items: center
    padding: 10px
    &-code
        background: #E7EAF4
        width: fit-content
        border-radius: 10px
    &-input
        width: 90%
        height: 150px
        border: 1px solid #7d59cc
        margin: 10px
        box-sizing: border-box
    &-button
        background: #FFFFFF
        border: 3px solid #B5ABE7
        border-radius: 10px
        width: 100px
        height: 30px
        color: #71588B
        cursor: pointer
        &:hover
            background: #B5ABE7
            color: #fff

.results
    display: flex
    flex-direction: column
    padding: 10px
    &-topic
        display: flex
        width: 100%
        flex-direction: column
        margin: 10px
        &-important
            background: #E7EAF4
            width: fit-content
            border-radius: 10px
            padding: 5px

.drops
    &-container
        display: flex
        flex-direction: column
        justify-content: center
        align-items: center
        max-height: 300px
        overflow: auto
    &-section
        width: 50%
        background: #E7EAF4
        margin: 10px
        &-item
            margin-top: 10px

.error
    position: absolute
    top: 0
    width: 200px
    height: 70px
    background: #CB2560
    border-radius: 10px
    margin-top: 20px
    display: flex
    justify-content: center
    align-items: center
    color: #FFF


