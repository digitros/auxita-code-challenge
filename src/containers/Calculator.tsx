import { ContactSupportOutlined } from '@material-ui/icons';
import React, { useEffect, useState } from 'react';
import Box from '../components/Box';

import CALCTYPES, {
  demoKidneyString,
  demoPressureString,
} from '../constants/CalculatorType';
import { BoxSizes } from '../interfaces/Props';

import '../styles/Calculator.styl';

interface IPressureDataObject {
  SysBP: number;
  DiaBP: number;
  atDate: string;
}

interface IKidneyDataObject {
  eGFR: number;
  atDate: string;
}

interface IDrops {
  drop: number;
  readings: IKidneyDataObject[];
}

type IDataObject = IPressureDataObject | IKidneyDataObject;

interface ICalculatorProps {
  type: string;
}

interface IError {
  message: string;
}

const isPressure = (param: IDataObject): param is IPressureDataObject => {
  return (param as IPressureDataObject).DiaBP !== undefined;
};

const Calculator = ({ type }: ICalculatorProps) => {
  const [dataInput, setDataInput] = useState('');
  const [dataObject, setDataObject] = useState<IDataObject[]>([]);
  const [error, setError] = useState<IError>({} as IError);
  const [lastReading, setLastReading] = useState<IDataObject>();
  const [classification, setClassification] = useState('');
  const [drops, setDrops] = useState<IDrops[]>([]);

  const parseStringToJSON = (stringToParse: string) => {
    try {
      setDataObject(JSON.parse(stringToParse));
    } catch (error) {
      setError({ message: 'Incorrect Format' });
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setDataInput(e.target.value);
  };

  const handleClick = () => {
    parseStringToJSON(dataInput);
  };

  const calculateHealth = (data: IDataObject[]) => {
    const calculateDisease = (reading: IDataObject) => {
      if (type === CALCTYPES.PRESSURE && isPressure(reading)) {
        if (reading.SysBP >= 180 && reading.DiaBP >= 120) {
          setClassification('Stage 3');
        } else if (
          reading.SysBP >= 160 &&
          reading.SysBP < 180 &&
          reading.DiaBP >= 100 &&
          reading.DiaBP < 110
        ) {
          setClassification('Stage 2');
        } else if (
          reading.SysBP >= 140 &&
          reading.SysBP < 160 &&
          reading.DiaBP >= 90 &&
          reading.DiaBP < 100
        ) {
          setClassification('Stage 1');
        } else {
          setClassification('No Hypertension');
        }
      } else if (type === CALCTYPES.KIDNEY && !isPressure(reading)) {
        if (reading.eGFR >= 90) {
          setClassification('Normal');
        } else if (reading.eGFR >= 60 && reading.eGFR <= 89) {
          setClassification('Mildy Decreased');
        } else if (reading.eGFR >= 45 && reading.eGFR <= 59) {
          setClassification('Mild to Moderate');
        } else if (reading.eGFR >= 30 && reading.eGFR <= 44) {
          setClassification('Moderate to Severe');
        } else if (reading.eGFR >= 15 && reading.eGFR <= 29) {
          setClassification('Deverely Decreased');
        } else {
          setClassification('Kidney Failure');
        }
      }
    };

    const calculateDrops = (data: IKidneyDataObject[]) => {
      const orderedData = data.sort((a, b) => a.atDate.localeCompare(b.atDate));
      const dropReadings = orderedData
        .map((item, index, arr) => {
          if (index !== arr.length - 1) {
            const dropPercentage = Math.floor(
              Math.abs(100 - (arr[index + 1].eGFR * 100) / item.eGFR)
            );
            if (dropPercentage >= 20) {
              return { drop: dropPercentage, readings: [item, arr[index + 1]] };
            }
          }
        })
        .filter((item) => item);
      setDrops(dropReadings as IDrops[]);
    };

    const isFormatCorrect = (dataFormat: IDataObject[]) => {
      try {
        if (type === CALCTYPES.PRESSURE) {
          return !dataFormat.some((item) => !isPressure(item));
        }
        return !dataFormat.some((item) => isPressure(item));
      } catch (error) {
        setError({ message: 'Incorrect Format' });
      }
    };

    if (isFormatCorrect(data)) {
      const last = data.sort((a, b) => a.atDate.localeCompare(b.atDate))[
        data.length - 1
      ];
      type === CALCTYPES.KIDNEY && calculateDrops(data as IKidneyDataObject[]);
      last && setLastReading(last);
      last && calculateDisease(last);
    } else {
      setError({ message: 'Incorrect Format' });
    }
  };

  useEffect(() => {
    dataObject && type === CALCTYPES.PRESSURE
      ? calculateHealth(dataObject as IPressureDataObject[])
      : calculateHealth(dataObject as IKidneyDataObject[]);
  }, [dataObject]);

  useEffect(() => {
    error.message && setTimeout(() => setError({} as IError), 5000);
  }, [error]);

  return (
    <main className="tools">
      {error.message && <div className="error">{error.message}</div>}
      <Box size={BoxSizes.large}>
        <div className="tools-title">
          <h2>{type}</h2>
        </div>
      </Box>
      <section className="tools-main">
        <Box size={BoxSizes.medium}>
          <div className="results">
            {lastReading ? (
              <>
                <div className="results-topic">
                  <h2>Last Reading:</h2>
                  {type === CALCTYPES.PRESSURE && isPressure(lastReading) && (
                    <span>
                      SysBP: <strong>{lastReading.SysBP}</strong> DisBP:{' '}
                      <strong>{lastReading.DiaBP}</strong>
                    </span>
                  )}
                  {type === CALCTYPES.KIDNEY && !isPressure(lastReading) && (
                    <span className="results-topic-important">
                      eGFR: <strong>{lastReading.eGFR}</strong>
                    </span>
                  )}
                </div>
                <span className="results-topic">
                  Last Date: <strong>{lastReading.atDate}</strong>
                </span>
                <span className="results-topic">
                  Classification: <strong>{classification}</strong>
                </span>
              </>
            ) : (
              <span className="results-topic">
                Here you will see your health calculation.
              </span>
            )}
            {drops.length > 0 && (
              <div className="drops-container">
                <h2>Important Drops</h2>
                {drops.map((drop, index) => {
                  return (
                    <div className="drops-section" key={`drop${index}`}>
                      <span>Drop Percentage: {drop.drop}%</span>
                      {drop.readings.map((read, readIndex) => {
                        return (
                          <div
                            key={`dropStat${readIndex}`}
                            className="drops-section-item"
                          >
                            <span>
                              eGFR: <strong>{read.eGFR}</strong>
                            </span>
                            <span>
                              Date: <strong>{read.atDate}</strong>
                            </span>
                          </div>
                        );
                      })}
                    </div>
                  );
                })}
              </div>
            )}
          </div>
        </Box>
        <Box size={BoxSizes.medium}>
          <div className="calculator">
            <span>Please write/paste your data in following format: </span>
            <pre className="calculator-code">
              {type === CALCTYPES.PRESSURE
                ? demoPressureString
                : demoKidneyString}
            </pre>
            <textarea className="calculator-input" onChange={handleChange} />
            <button className="calculator-button" onClick={handleClick}>
              Send
            </button>
          </div>
        </Box>
      </section>
    </main>
  );
};

export default Calculator;
