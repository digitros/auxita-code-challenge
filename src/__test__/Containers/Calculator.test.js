import React from 'react';
import {create} from 'react-test-renderer'

/* Containers */
import Calculator from '../../containers/Calculator';

/* Constants */
import CALCTYPE from '../../constants/CalculatorType';

jest.mock('../../i18n.js');

describe('<Calculator />', () => {
  test('Calculator snapshot', () => {
    const calculator = create(<Calculator type={CALCTYPE.KIDNEY} />);
    expect(calculator.toJSON()).toMatchSnapshot();
  });
});
