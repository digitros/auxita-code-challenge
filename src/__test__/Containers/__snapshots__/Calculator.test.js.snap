// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator /> Calculator snapshot 1`] = `
<main
  className="tools"
>
  <section
    className="box box-large"
  >
    <div
      className="tools-title"
    >
      <h2>
        Nav:Kidney
      </h2>
    </div>
  </section>
  <section
    className="tools-main"
  >
    <section
      className="box box-medium"
    >
      <div
        className="results"
      >
        <span
          className="results-topic"
        >
          Calculator:BeforeCalc
        </span>
      </div>
    </section>
    <section
      className="box box-medium"
    >
      <div
        className="calculator"
      >
        <span>
          Calculator:Input
          :
           
        </span>
        <pre
          className="calculator-code"
        >
          
  [
    {"eGFR": 30, "atDate": "2018/10/31"},
    {"eGFR": 70, "atDate": "2018/10/20"},
    {"eGFR": 62, "atDate": "2017/10/20"}
  ]
    
        </pre>
        <textarea
          className="calculator-input"
          onChange={[Function]}
        />
        <button
          className="calculator-button"
          onClick={[Function]}
        >
          Calculator:Send
        </button>
      </div>
    </section>
  </section>
</main>
`;

exports[`<Calculator /> snapshot Calculator component render 1`] = `
<main
  className="tools"
>
  <section
    className="box box-large"
  >
    <div
      className="tools-title"
    >
      <h2>
        Nav:Kidney
      </h2>
    </div>
  </section>
  <section
    className="tools-main"
  >
    <section
      className="box box-medium"
    >
      <div
        className="results"
      >
        <span
          className="results-topic"
        >
          Calculator:BeforeCalc
        </span>
      </div>
    </section>
    <section
      className="box box-medium"
    >
      <div
        className="calculator"
      >
        <span>
          Calculator:Input
          :
           
        </span>
        <pre
          className="calculator-code"
        >
          
  [
    {"eGFR": 30, "atDate": "2018/10/31"},
    {"eGFR": 70, "atDate": "2018/10/20"},
    {"eGFR": 62, "atDate": "2017/10/20"}
  ]
    
        </pre>
        <textarea
          className="calculator-input"
          onChange={[Function]}
        />
        <button
          className="calculator-button"
          onClick={[Function]}
        >
          Calculator:Send
        </button>
      </div>
    </section>
  </section>
</main>
`;
